### ISSUE TITLE
Confusing unchecked dependency issue

### ISSUE THREAD

**Issue by devonestes** [elixir-lang/elixir#10126]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
Erlang/OTP 22 [erts-10.7.2.1] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [hipe]

Elixir 1.11.0-dev (e51cc8b) (compiled with Erlang/OTP 21)
* Operating system: 
MacOS 10.15.5 (19F101)

### Current behavior
I just installed Elixir `master` with `asdf` to try out the compilation improvements on my application,
but when I ran `mix compile` I saw this output:

```
Unchecked dependencies for environment dev:
* absinthe_phoenix (Hex package)
  the dependency does not match the requirement "~> 1.5.0-rc0", got "1.5.0-rc.0"
** (Mix) Can't continue due to errors on dependencies
```

This application compiles and runs just fine on Elixir 1.10.3.

I then ran `mix deps.get` to resolve this, and it upgraded my to these dependencies:
```
  absinthe 1.5.0-rc.2 => 1.5.1
  absinthe_phoenix 1.5.0-rc.0 => 1.5.0
  absinthe_plug 1.5.0-rc.1 => 1.5.0
```

### Expected behavior

Given that this `mix.lock` file and these dependencies work just fine on 1.10 I would have
expected them to work on 1.11, but maybe there is a breaking change somewhere that I
missed about resolving dependencies?

**Comment by josevalim** [comment]
Can you please provide a sample application that reproduces the error? It looks like a regression. Thanks.

**Comment by devonestes** [comment]
Here's a download link to a zipped project folder that reproduces the error: https://send.firefox.com/download/f51daac5b6ca19cd/#m5XKltuREI3n_uYd7dzYDg. The exact deps in there are sort of important, since they're the ones that are currently being used successfully on 1.10.3, and so when I switch Elixir versions with these deps already downloaded it fails.

The repro steps are:

1) Unzip this
2) run `mix compile` with the locked Elixir version in `asdf`(and see failure)
3) run `asdf local elixir 1.10.3-otp-22`
4) run `mix compile` (and see success)

**Comment by ericmj** [comment]
This was a bug fixed in 1.11.0. You have requirement `~> 1.5.0-rc0` and match against version `1.5.0-rc.0`, but according to semver `1.5.0-rc0 > 1.5.0-rc.0` so the requirement should not match an earlier version.

**Comment by josevalim** [comment]
oh, the `.`! it would have taken me forever to find it. Thanks @ericmj!
